core/domain/classifier_services_test.py

```diff
@@ -623,3 +630,162 @@ def test_can_not_store_classifier_data_due_to_invalid_job_id(self):
+                'algorithm_id': 'NewTextClassifier',
+                'algorithm_version': 1
+                },
+            }
+        with self.swap(
+            feconf,
+            'INTERACTION_CLASSIFIER_MAPPING',
+            mock_interaction_classifier_mapping):
+            classifier_services.migrate_state_training_jobs(
+                classifier_services.get_state_training_jobs_mapping(
+                    self.exp_id, 1, state_name)
+             )
+        state_training_jobs_mapping = (
+            classifier_services.get_state_training_jobs_mapping(
+                self.exp_id, 1, 'Home')
+            )
+        self.assertIn(
+            'NewTextClassifier',
+            state_training_jobs_mapping.algorithm_ids_to_job_ids)
+
+        # Check migration of an existing algorithm to a newer version.
+        mock_interaction_classifier_mapping = {
+            'TextInput': {
+                'algorithm_id': 'NewTextClassifier',
+                'algorithm_version': 2
+                },
+            }
+        with self.swap(
+            feconf,
+            'INTERACTION_CLASSIFIER_MAPPING',
+            mock_interaction_classifier_mapping):
+            classifier_services.migrate_state_training_jobs(
+                classifier_services.get_state_training_jobs_mapping(
+                    self.exp_id, 1, state_name))
+        next_job = classifier_services.fetch_next_job()
+        self.assertEqual(
+            mock_interaction_classifier_mapping[
+                'TextInput']['algorithm_version'],
+            next_job.algorithm_version
+        )
+
+    def test_reverted_exploration_maintains_classifier_model_mapping(self):
+        """Test if the classifier model mapping is maintained when an
+        exploration is reverted.
+        """
+
+        state_name = 'Home'
+        old_exploration = exp_fetchers.get_exploration_by_id(self.exp_id)
+        interaction_id = old_exploration.states[
+            state_name].interaction.id
+        algorithm_id = feconf.INTERACTION_CLASSIFIER_MAPPING[
+            interaction_id]['algorithm_id']
+        job = classifier_services.get_classifier_training_job(
+            self.exp_id, old_exploration.version, state_name,
+            algorithm_id)
+        change_list = [exp_domain.ExplorationChange({
+            'cmd': exp_domain.CMD_EDIT_EXPLORATION_PROPERTY,
+            'property_name': 'title',
+            'new_value': 'A new title'
+        })]
+        with self.swap(feconf, 'ENABLE_ML_CLASSIFIERS', True):
+            exp_services.update_exploration(
+                feconf.SYSTEM_COMMITTER_ID, self.exp_id, change_list, '')
+        current_exploration = exp_fetchers.get_exploration_by_id(self.exp_id)
+        (
+            classifier_services
+            .create_classifier_training_job_for_reverted_exploration(
+            current_exploration, old_exploration))
+        # Verify if classifier model mapping is maintained using the job ID.
```

Comment:
I'm afraid I don't really understand what you're testing here, just from reading the test name and comments. What has the job ID got to do with the "classifier model mapping"? They seem like unrelated concepts...

(To put it another way I'm not sure what a "wrong" result for this test case would look like.)

Reply:
@seanlip I will try to explain what the test is doing. From what I have understood, upon reverting an exploration to an older version, `create_classifier_training_job_for_reverted_exploration` will copy the mappings from the old version of the exploration (one which we would be reverting to) to a new version whose version number would be 1 more than the current exploration version.
So what I am doing is to store the job ID of a classifier job for a given state from the old version, make changes to the exploration and call the `create_classifier_training_job_for_reverted_exploration` which would be called when an exploration is reverted. I then retrieve the job ID for the same state for the reverted exploration version and check if it matches the job ID for the old version. This would fail if the `create_classifier_training_job_for_reverted_exploration` fails to maintain the classifier model mapping.

Reply:
@JayVivarekar  Thanks for the explanation. The thing I don't understand is why you're going with this approach, which seems a bit convoluted. Why not instead: create an exploration, make changes, store the job ID, revert the exploration to the older version, and then check the job ID again?